File: backend/app/services/jd_analyzer.py
"""Job description analyzer."""

import re

from app.schemas.schemas import JobDescriptionAnalysis
from app.services.nlp_engine import extract_keywords_tfidf, extract_skills_from_text
from app.utils.text_utils import dedupe_preserve_order, extract_sentences, normalize_text


def _extract_list_items(text: str, headers: list[str]) -> list[str]:
  items: list[str] = []
  pattern = "|".join(re.escape(h) for h in headers)
  sections = re.split(rf"(?i)(?:{pattern})\s*:?\s*\n", text)
  for section in sections[1:]:
    for line in section.split("\n"):
      line = line.strip()
      if re.match(r"^[\s•\-\*]", line):
        items.append(re.sub(r"^[\s•\-\*]+\s*", "", line))
      elif line and len(line) > 15 and not re.match(r"^[A-Z\s]{3,}$", line):
        if len(items) < 20:
          items.append(line)
    break
  return items[:15]


def analyze_job_description(text: str) -> JobDescriptionAnalysis:
  """Extract skills, requirements, and keywords from a job description."""
  text = normalize_text(text)
  technical, soft = extract_skills_from_text(text)

  responsibilities = _extract_list_items(
    text,
    ["responsibilities", "duties", "what you will do", "what you'll do", "role"],
  )
  requirements = _extract_list_items(
    text,
    ["requirements", "qualifications", "must have", "required", "minimum qualifications"],
  )

  if not responsibilities:
    responsibilities = [
      s for s in extract_sentences(text)
      if any(w in s.lower() for w in ("responsible", "develop", "manage", "lead", "design", "implement"))
    ][:8]

  if not requirements:
    requirements = [
      s for s in extract_sentences(text)
      if any(w in s.lower() for w in ("required", "must", "minimum", "years", "degree", "experience"))
    ][:8]

  keywords = extract_keywords_tfidf([text], top_n=40)
  # High priority = appears in requirements + technical skills
  req_text = " ".join(requirements).lower()
  high_priority = dedupe_preserve_order([
    kw for kw in keywords
    if kw.lower() in req_text or kw.lower() in " ".join(technical).lower()
  ][:20])

  # Supplement keywords with extracted skills
  all_keywords = dedupe_preserve_order(keywords + technical + soft)

  return JobDescriptionAnalysis(
    raw_text=text,
    technical_skills=technical,
    soft_skills=soft,
    responsibilities=responsibilities,
    requirements=requirements,
    keywords=all_keywords,
    high_priority_keywords=high_priority or all_keywords[:15],
  )

File: frontend/src/engine/jdAnalyzer.js
import { extractKeywordsTfidf, extractSkillsFromText } from './nlpEngine'
import { dedupePreserveOrder, extractSentences, normalizeText } from './textUtils'

function extractListItems(text, headers) {
  const pattern = headers.map((h) => h.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')).join('|')
  const sections = text.split(new RegExp(`(?:${pattern})\\s*:?\\s*\\n`, 'i'))
  const items = []
  if (sections.length > 1) {
    for (const line of sections[1].split('\n')) {
      const trimmed = line.trim()
      if (/^[\s•\-*]/.test(trimmed)) items.push(trimmed.replace(/^[\s•\-*]+\s*/, ''))
      else if (trimmed.length > 15 && !/^[A-Z\s]{3,}$/.test(trimmed) && items.length < 20) items.push(trimmed)
    }
  }
  return items.slice(0, 15)
}

export function analyzeJobDescription(text) {
  const raw_text = normalizeText(text)
  const [technical_skills, soft_skills] = extractSkillsFromText(raw_text)

  let responsibilities = extractListItems(raw_text, ['responsibilities', 'duties', 'what you will do', "what you'll do", 'role'])
  let requirements = extractListItems(raw_text, ['requirements', 'qualifications', 'must have', 'required', 'minimum qualifications'])

  if (!responsibilities.length) {
    responsibilities = extractSentences(raw_text).filter((s) =>
      /responsible|develop|manage|lead|design|implement/i.test(s),
    ).slice(0, 8)
  }
  if (!requirements.length) {
    requirements = extractSentences(raw_text).filter((s) =>
      /required|must|minimum|years|degree|experience/i.test(s),
    ).slice(0, 8)
  }

  const keywords = extractKeywordsTfidf([raw_text], 40)
  const reqText = requirements.join(' ').toLowerCase()
  const techText = technical_skills.join(' ').toLowerCase()
  const high_priority_keywords = dedupePreserveOrder(
    keywords.filter((kw) => reqText.includes(kw.toLowerCase()) || techText.includes(kw.toLowerCase())),
  ).slice(0, 20)

  const allKeywords = dedupePreserveOrder([...keywords, ...technical_skills, ...soft_skills])

  return {
    raw_text,
    technical_skills,
    soft_skills,
    responsibilities,
    requirements,
    keywords: allKeywords,
    high_priority_keywords: high_priority_keywords.length ? high_priority_keywords : allKeywords.slice(0, 15),
  }
}
